### ISSUE TITLE
Element queries integration

### ISSUE THREAD

**Issue by pluma** [paypal/glamorous#182]
As discussed on gitter, I think glamor(ous) would be a great match for element queries (via [eqcss](http://elementqueries.com)). I'm not sure whether this is more about glamor than glamorous (upon further consideration I'm thinking it's likely the former) but since I mentioned it on the glamorous gitter, I'm filing it here for discussion.

---

In a nutshell, element queries provide two things on top of regular CSS:

1. They allow writing conditions that match against specific properties of a single element (foregoing the need for JS for behaviour like "hide this element if the container is smaller than N px"); but with [a lot more checks than just width and height](http://elementqueries.com#element-query-conditions).

2. Since they behave similarly to media queries, they can define arbitrary CSS rules that only take effect when the element query matches (just like media queries which can define CSS rules that only take effect when the media query matches).

Additionally element queries define semantics for referencing the element matched by the query in selectors.

The former is pretty useful to reduce DOM boilerplate by itself, but the latter is the real killer because it means you can write CSS like (stupid example, I know) this:

```css
@element .cls-xyz {
  body {
    background-color: red;
  }
}
```

(in plain English: "if an element with the class name `cls-xyz` is present on the page, set the background-color of the body element to red.)

This eliminates a lot of headaches for all kinds of cross-cutting styling concerns that involve elements that are neither direct siblings nor descendants of one another.

---

The [EQCSS library](https://yarnpkg.com/en/package/eqcss) basically acts as a runtime that applies the element queries by watching for changes to the DOM. It can either parse the existing stylesheets by calling `EQCSS.load()` or can be fed new element queries at runtime like this (example from [eqcss-parse](https://yarnpkg.com/en/package/eqcss-parse)):

```js
EQCSS.register([{
  selector: 'input',
  conditions: [{
    measure: 'max-characters',
    value: '3',
    unit: null
  }],
  style: 'html {background: lime;}'
}]);
```

So implementing element queries in glamor(ous) should be as easy as intercepting element queries and passing them to `EQCSS.register` instead of dumping them in the stylesheet the generated CSS would normally be written to.

I think it would be possible to build a proof of concept by patching glamor(ous) directly but I'm drowning in work at the moment so I have to leave this as an exercise for the reader right now.

---

Is this at all in scope for glamor(ous)? I think it suits the idea of css-in-js nicely (even though EQCSS often feels more like js-in-css).

/cc @tomhodgins

**Comment by tomhodgins** [comment]
Feel free to ask me anything about the EQCSS-side of this equation, I'm less familiar with Glamorous so I can only see how the connection could be made from the EQCSS direction. If Glamorous can produce equivalent JS objects in the same format that `eqcss-parse` returns then you may not even need to use EQCSS syntax at all.

Assuming you're working with CSS in JS and eventually that CSS will get output to the page, here are the possible handoff points I can see:

1. your JS directly produces objects `EQCSS.js` can load as JS (same format output as `eqcss-parse`)
2. your JS includes EQCSS syntax strings that can be parsed in advance and loaded as JS
3. your JS outputs EQCSS syntax to HTML (as style, link, or script) to be found and parsed by `EQCSS.js` after the page loads

To me it sounds like either 1. or 2. are what you're aiming for here

**Comment by kentcdodds** [comment]
Actually, this feels like a problem that would be solved by glamor more than glamorous. Maybe @threepointone, @donaldpipowitch, or @ChristopherBiscardi could pitch in with thoughts.

**Comment by ChristopherBiscardi** [comment]
maybe something worth exploring in a [plugin](https://github.com/threepointone/glamor/blob/master/docs/plugins.md)?

**Comment by pluma** [comment]
@ChristopherBiscardi If I understand it correctly, a possible plugin would be to do something like this?

```js
function eqcss({selector, style}) {
  if (selector.startsWith('@element ')) {
    EQCSS.process(`${selector} { ${style} }`);
  }
  return {selector, style};
}
```

This seems straightforward enough but if I understand the plugin system correctly the `style` wouldn't benefit from things like auto-prefixing?

**Comment by kentcdodds** [comment]
Hmmm... I _think_ you'd still get that, but I'm not sure 🤔 Maybe try it out: https://help.glamorous.rocks

**Comment by pluma** [comment]
Here's a proof of concept:

https://codesandbox.io/s/g5ZgWRGPk

Note that the contained style doesn't get processed by glamor, though. I guess they'd need to be run through the entire plugin stack manually again (which shouldn't cause any trouble but I'm not sure whether there's an API for that).

**Comment by kentcdodds** [comment]
Cool! Someone who knows glamor plugins better than I do could probably help with the autoprefixing stuff I imagine 🤔 

At this point I think that I'm going to close this issue (as it seems out of scope for glamorous), but feel free to keep chatting about a solution. I hope that we can make this plugin a package that people can use and we could add it to the glamorous website under "Integrations" :)
